### PR TITLE
Wip

### PR DESCRIPTION
**Pull request by pablodiego**

### CODE REVIEW

**Review by lucasccordeiro — CHANGES_REQUESTED** [review]
Shall we add some test cases to check these methods?

File: src/util/mp_arith.cpp [hunk 1]
```diff
@@ -150,3 +153,142 @@ unsigned integer2unsigned(const BigInt &n)
...
+    tmp_a /= 2;
+    tmp_b /= 2;
+  }
+
+  return result;
+}
+
+BigInt bitwise_or(const BigInt &a, const BigInt &b)
+{
+  (!a.is_negative() && !b.is_negative());
```
Comment (rafaelsamenezes) [inline src/util/mp_arith.cpp:181]: What is this doing?

File: src/util/mp_arith.cpp [hunk 2]
```diff
@@ -150,3 +153,142 @@ unsigned integer2unsigned(const BigInt &n)
...
+
+  if(a.is_ulong() && b.is_ulong())
+    return a.to_ulong() | b.to_ulong();
+
+  return bitwise(a, b, [](bool a, bool b) { return a || b; });
+}
+
+BigInt bitwise_and(const BigInt &a, const BigInt &b)
+{
+  (!a.is_negative() && !b.is_negative());
```
Comment (rafaelsamenezes) [inline src/util/mp_arith.cpp:191]: What is this doing?

File: src/util/mp_arith.cpp [hunk 3]
```diff
@@ -150,3 +153,142 @@ unsigned integer2unsigned(const BigInt &n)
...
+
+  if(a.is_ulong() && b.is_ulong())
+    return a.to_ulong() & b.to_ulong();
+
+  return bitwise(a, b, [](bool a, bool b) { return a && b; });
+}
+
+BigInt bitwise_xor(const BigInt &a, const BigInt &b)
+{
+  (!a.is_negative() && !b.is_negative());
```
Comment (rafaelsamenezes) [inline src/util/mp_arith.cpp:201]: What is this doing?

**Comment by rafaelsamenezes** [comment]
> Shall we add some test cases to check these methods?

Yes, bitwise operations should be added inside [math suite](https://github.com/esbmc/esbmc/blob/master/unit/big-int/bigint.test.cpp#L349) or in a newer suite. Also, it would be great if you could add [fuzzing tests](https://github.com/esbmc/esbmc/blob/master/unit/big-int/bigint.fuzz.cpp)

File: src/util/mp_arith.cpp (hunk 1 above)
Comment (mikhailramalho) [reply src/util/mp_arith.cpp:181]: In CBMC, these are the arguments of a macro called PRECONDITION. 

In ESBMC, they should be assertions.

**Comment by mikhailramalho** [comment]
> > Shall we add some test cases to check these methods?
> 
> Yes, bitwise operations should be added inside [math suite](https://github.com/esbmc/esbmc/blob/master/unit/big-int/bigint.test.cpp#L349) or in a newer suite. Also, it would be great if you could add [fuzzing tests](https://github.com/esbmc/esbmc/blob/master/unit/big-int/bigint.fuzz.cpp)

I agree with adding fuzzing tests because they are not used anywhere in ESBMC. In the future, they should be used to simplify expressions in `simplify_expr2.cpp`.

**Comment by lucasccordeiro** [comment]
@mikhailramalho: May we close this PR?

**Comment by mikhailramalho** [comment]
> @mikhailramalho: May we close this PR?

Yes.

**Pull request closed without merging**